src/mame/video/mcd212.cpp

```diff
@@ -1231,97 +1153,222 @@ uint8_t mcd212_device::csr2_r()
+			switch (mixing_mode & 7)
+			{
+				case 0: // No Mosaic A/B, A->B->Backdrop plane ordering
+					mix_lines<false, false, true>(plane_a, transparent_a, plane_b, transparent_b, out);
+					break;
+				case 1: // Mosaic A, No Mosaic B, A->B->Backdrop plane ordering
+					mix_lines<true, false, true>(plane_a, transparent_a, plane_b, transparent_b, out);
+					break;
+				case 2: // No Mosaic A, Mosaic B, A->B->Backdrop plane ordering
+					mix_lines<false, true, true>(plane_a, transparent_a, plane_b, transparent_b, out);
+					break;
+				case 3: // Mosaic A/B, A->B->Backdrop plane ordering
+					mix_lines<true, true, true>(plane_a, transparent_a, plane_b, transparent_b, out);
+					break;
+				case 4: // No Mosaic A/B, B->A->Backdrop plane ordering
+					mix_lines<false, false, false>(plane_a, transparent_a, plane_b, transparent_b, out);
+					break;
+				case 5: // Mosaic A, No Mosaic B, B->A->Backdrop plane ordering
+					mix_lines<true, false, false>(plane_a, transparent_a, plane_b, transparent_b, out);
+					break;
+				case 6: // No Mosaic A, Mosaic B, B->A->Backdrop plane ordering
+					mix_lines<false, true, false>(plane_a, transparent_a, plane_b, transparent_b, out);
+					break;
+				case 7: // Mosaic A/B, B->A->Backdrop plane ordering
+					mix_lines<true, true, false>(plane_a, transparent_a, plane_b, transparent_b, out);
+					break;
+			}
+
+			draw_cursor(out);
+		}
 	}
 
+	// Toggle frame parity at the end of the visible frame (even in non-interlaced mode).
 	if (scanline == (m_total_height - 1))
 	{
-		m_csrr[0] ^= 0x20;
+		m_csrr[0] ^= CSR1R_PA;
 	}
 
 	return 0;
 }
 
+template int mcd212_device::ram_dtack_cycle_count<0>();
+template int mcd212_device::ram_dtack_cycle_count<1>();
+
+template <int Channel>
+int mcd212_device::ram_dtack_cycle_count()
+{
+	// Per MCD-212 documentation, it takes 4 CLKs (2 SCC68070 clocks) for a VRAM access during the System timing slot.
+
+	// No contending for Ch.1/Ch.2 timing slots if display is disabled
+	if (!BIT(m_dcr[0], DCR_DE_BIT))
+		return 2;
+
+	// No contending for Ch.1/Ch.2 timing slots if a relevant channel is disabled
+	if (!BIT(m_dcr[Channel], DCR_ICA_BIT))
+		return 2;
+
+	const int x = screen().hpos();
+	const int y = screen().vpos();
+	const bool x_outside_active_display = (x >= 408);
+
+	// No contending for Ch.1/Ch.2 timing slots during the final 8-pixel area on all lines
+	if (x >= 472)
+		return 2;
+
+	// No contending for Ch.1/Ch.2 timing slots during the free-run area of ICA lines
+	if (y < m_ica_height && x_outside_active_display)
+		return 2;
+
+	// No contending for Ch.1/Ch.2 timing slots during the free-run area of DCA lines if DCA is disabled
+	if (!BIT(m_dcr[Channel], DCR_DCA_BIT) && x_outside_active_display)
+		return 2;
+
+	// System access is restricted to the last 5 out of every 16 CLKs.
+	const int slot_cycle = (int)(machine().time().as_ticks(clock()) & 0xf);
+	if (slot_cycle >= 11)
+		return 2;
```

Comment:
Can you call `total_cycles()` on the CPU rather than doing operations on attotimes?

Reply:
No. The MCD-212 operates a 30MHz clock. The SCC68070 operates on a 15MHz clock after cycles-to-clocks and clocks-to-cycles is accounted for. Although that lack of granularity inevitably crops up in the event that the SCC68070 is doing an access which hits the 11 30MHz clock Channel-exclusive window, I'd much rather not have the window comparison itself be inaccurate as well.

If your concern is one of performance, I'd welcome you to do a before/after test with any CD-i game you choose to name. The difference is negligible if detectable at all.

Reply:
It wasn‘t so much performance I was thinking about, but skew.  MAME doesn’t always do very well with clocks that should be integer multiples of each other.

Reply:
All I can say with certainty is that the change has no effect on almost every game that I tested, and only positive results on the games that were affected, so clock skew would not have been an issue in the good handful of games that were tested.

Beyond calling `total_cycles()` requiring the VDSC's slice to effectively be narrowed to 10 clock cycles or widened to 12, there's a cleanliness issue as well - the VDSC has no access to the main CPU currently, and it would seem to go against the grain of device isolation to make it aware of the main CPU solely for this one function.

The CPU requests access to both the ROM and video RAM via the VDSC, so the current unidirectional dependency at least makes sense. Making the VDSC depend on the CPU doesn't, and having the handler pass the result of `total_cycles()` into the relevant delay-fetch functions seems about as ugly as it is now.

Reply:
Sure, I’m not saying it needs to change, and even if it does skew, it’s unlikely to cause issues.  Put it down as one of those things we need a better solution to at some point in the future.

Reply:
Sure, agreed. I'm not saying your suggestions are necessarily wrong, just providing the necessary context as to why they might not be the best solution in this case.

Reply:
In any event, is there anything I need to address in this PR before it goes in?